### PR TITLE
feat(machines): display the count of selected machines

### PR DESCRIPTION
**Pull request by huwshimi**
## Done

- Display the count of machines that have been selected in the machine list.

### MAAS deployment

To run this branch you will need access to one of the following MAAS deployments:

- [Karura](/HACKING.md#karura)
- [Bolla](/HACKING.md#bolla)
- [A development MAAS](/HACKING.md#development-deployment)
- [A local snap MAAS](/HACKING.md#snap-deployment) (this will not usually have machines)

### Running the branch

You can run this branch by:

- Serving with [dotrun](/HACKING.md#maas-ui-development-setup)
- [Building in a development MAAS](/HACKING.md#running-maas-ui-from-a-development-maas)

### QA steps

- Go to the machine list.
- Tick some machines and you should see the count update instantly in the header.
- Untick all machines and tick a group (you'll need to be grouped by something other than status e.g. owner). You should see the count get loaded and appear in the header (note that this number doesn't currently match the number of machines in the group - this is being fixed on the backend).
- Now tick some machines as well as the group. You should see the count update.
- Now tick the checkbox in the table header and "All machines selected" should appear.

## Fixes

Fixes: canonical/app-tribe#1102.

### CODE REVIEW

**Comment by huwshimi** [comment]
@amylily1011 @rayito any thoughts about what we should display when the count of selected machines is being updated? Depending on the MAAS it could be really quick or really slow. At the moment it changes to a spinner but on a fast MAAS it looks a bit jumpy. Not sure if that's OK or if we should move the spinner next to the existing count while it updates? Any suggestions?

Screen Recording 2022-09-08 at 1 38 35 pm

**Comment by amylily1011** [comment]
@rayito and I think this is ok. Making changes to this interaction won't be a big impact. So I think we can keep it this way :D

( cc @huwshimi )

File: src/app/machines/views/MachineList/MachineListHeader/MachineListHeader.test.tsx [hunk 1]
```diff
@@ -102,65 +110,144 @@ describe("MachineListHeader", () => {
...
+    state.machine.selectedMachines = {
+      items: ["abc123"],
+      groups: ["admin"],
+      grouping: FetchGroupKey.Owner,
+    };
+    state.machine.counts["mocked-nanoid-2"] = machineStateCountFactory({
+      count: 10,
+      loaded: true,
+    });
+    state.machine.counts["mocked-nanoid-3"] = machineStateCountFactory({
```
Comment (petermakowski) [inline src/app/machines/views/MachineList/MachineListHeader/MachineListHeader.test.tsx:215]: Might be worth using a variable as I suggested above and/or adding a comment explaining that `mocked-nanoid-3` refers to the count of machines in a selected group.

File: src/app/machines/views/MachineList/MachineListHeader/MachineListHeader.test.tsx [hunk 2]
```diff
@@ -102,65 +110,144 @@ describe("MachineListHeader", () => {
...
+  });
+
+  it("displays a selected count if some machines and groups have been selected", () => {
+    state.machine.selectedMachines = {
+      items: ["abc123"],
+      groups: ["admin"],
+      grouping: FetchGroupKey.Owner,
+    };
+    state.machine.counts["mocked-nanoid-2"] = machineStateCountFactory({
+      count: 10,
```
Comment (petermakowski) [inline src/app/machines/views/MachineList/MachineListHeader/MachineListHeader.test.tsx:212]:
```suggestion
      count: allMachinesCount,
```

File: src/app/machines/views/MachineList/MachineListHeader/MachineListHeader.test.tsx [hunk 3]
```diff
@@ -102,65 +110,144 @@ describe("MachineListHeader", () => {
...
+      loaded: true,
+    });
+    state.machine.counts["mocked-nanoid-3"] = machineStateCountFactory({
+      count: 2,
```
Comment (petermakowski) [inline src/app/machines/views/MachineList/MachineListHeader/MachineListHeader.test.tsx:216]:
```suggestion
      count: selectedMachinesCount,
```

File: src/app/machines/views/MachineList/MachineListHeader/MachineListHeader.tsx [hunk 4]
```diff
@@ -28,24 +29,56 @@ import { getNodeActionTitle } from "app/store/utils";
...
+  } = useFetchMachineCount(filters);
+  // Only add the count if there are filters as sending `null` filters
+  // to the count API will return a count of all machines.
+  if (filters) {
+    selectedCount += fetchedSelectedCount;
+  }
+  const onlyHasItems =
+    !!selectedMachines &&
+    selectedCount > 0 &&
+    (!("groups" in selectedMachines) || !selectedMachines?.groups?.length);
```
Comment (petermakowski) [inline src/app/machines/views/MachineList/MachineListHeader/MachineListHeader.tsx:81]: Do we have a use case for getting a count that matches both the groups and the items?

It seems like this should either already be a part of the `useFetchMachineCount` hook or abstracted to another one, e.g. `useSelectedMachineCount`, along with logic below - `machineCountLoading || onlyHasItems ? false : selectedLoading`).

**Review by petermakowski — APPROVED** [review]

**Comment by petermakowski** [comment]
Merging as I need these changes for other issues I'm working on.

**Pull request merged**